mislabeled/probe/_linear.py

```diff
@@ -85,63 +101,100 @@ def grad_y(self, X, y):
             raise NotImplementedError()
         return dl_dy
 
+    @property
+    def packed_coef(self):
+        packed = self.coef
+        if self.intercept is not None:
+            packed = np.concatenate((packed, self.intercept[None, :]))
+        return packed
+
+    @property
+    def packed_regul(self):
+        if self.regul is None:
+            return np.zeros_like(self.packed_coef)
+
+        packed = np.full_like(self.coef, self.regul)
+        if self.intercept is not None:
+            packed = np.concatenate((packed, np.zeros_like(self.intercept[None, :])))
+        return packed
+
     def grad_p(self, X, y):
         # gradients w.r.t. the parameters (weight, intercept)
         dl_dy = self.grad_y(X, y)
 
         X_p = X if not sp.issparse(X) else X.toarray()
-        if self.intercept is not None:
-            X_p = np.hstack((X_p, np.ones((X_p.shape[0], 1))))
+        X_p = self.add_bias(X_p)
 
-        return dl_dy[:, :, None] * X_p[:, None, :]
+        dl_dp = dl_dy[:, None, :] * X_p[:, :, None]
+        # dl_dp -= 2 * self.packed_regul * self.packed_coef / X.shape[0]
+        return dl_dp.reshape(X_p.shape[0], -1)
 
     def grad_X(self, X, y):
         # gradients w.r.t the input features
         dl_dy = self.grad_y(X, y)
         dy_dX = self.coef.T
         return dl_dy @ dy_dX
 
-    def hessian(self, X, y):
-        X_p = X if not sp.issparse(X) else X.toarray()
-
+    def add_bias(self, X):
         if self.intercept is not None:
-            X_p = np.hstack((X_p, np.ones((X_p.shape[0], 1))))
+            if sp.issparse(X):
+                return sp.hstack((X, np.ones((X.shape[0], 1))))
+            else:
+                return np.hstack((X, np.ones((X.shape[0], 1))))
+        return X
 
+    def pseudo(self, X):
+        if (k := self.out_dim) > 1:
+            if sp.issparse(X):
+                return sp.kron(X, sp.eye(k))
+            else:
+                return np.kron(X, np.eye(k))
+        return X
+
+    def variance(self, p):
```

Comment:
why it this called the "variance" ? Can you elaborate a bit in a docstring ?

Reply:
For me it's the variance of the link function from the according GLM model.